File: src/01_data_collection.R
library(quantmod)
library(tidyverse)
library(lubridate)
library(here)

options(scipen = 999) 
options(warnings = -1) 

# Configuration
symbols <- c("AAPL", "MSFT", "NVDA", "TSLA", "BTC-USD", "ETH-USD", "GC=F", "CL=F", "^GSPC", "^VIX")
start_date <- Sys.Date() - years(5)
end_date <- Sys.Date()

# Fonction de téléchargement robuste
download_stock_data <- function(symbol, start, end) {
  tryCatch({
    cat("Téléchargement de", symbol, "...\n")
    data <- getSymbols(symbol, 
                       src = "yahoo",
                       from = start,
                       to = end,
                       auto.assign = FALSE)
    
    # Nettoyer les noms de colonnes
    df <- as.data.frame(data) %>%
      rownames_to_column("Date") %>%
      mutate(Date = as.Date(Date),
             Symbol = symbol) %>%
      rename_all(~str_replace_all(., paste0(symbol, "."), ""))
    
    # Calculer les returns
    df <- df %>%
      mutate(Returns = (Close - lag(Close)) / lag(Close),
             Log_Returns = log(Close / lag(Close)))
    
    return(df)
  }, error = function(e) {
    cat("Erreur pour", symbol, ":", e$message, "\n")
    return(NULL)
  })
}

# Télécharger toutes les données
all_data <- map_dfr(symbols, ~download_stock_data(.x, start_date, end_date))

# Créer les dossiers s'ils n'existent pas
dir.create("data", showWarnings = FALSE)
dir.create("data/raw", showWarnings = FALSE)

# Sauvegarder
write_csv(all_data, here("data/raw/financial_data_raw.csv"))
saveRDS(all_data, here("data/raw/financial_data_raw.rds"))

# Statistiques descriptives
summary_stats <- all_data %>%
  group_by(Symbol) %>%
  summarise(
    Start = min(Date),
    End = max(Date),
    N_obs = n(),
    Missing = sum(is.na(Close)),
    Mean_Price = mean(Close, na.rm = TRUE),
    SD_Price = sd(Close, na.rm = TRUE),
    Mean_Volume = mean(Volume, na.rm = TRUE)
  )

write_csv(summary_stats, here("results/tables/data_summary.csv"))

File: src/01_data_collection.py
import yfinance as yf
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

def download_financial_data(symbols, period="5y"):
    """
    Télécharge les données financières pour une liste de symboles
    """
    all_data = []
    
    for symbol in symbols:
        try:
            print(f"Downloading {symbol}...")
            ticker = yf.Ticker(symbol)
            df = ticker.history(period=period)
            df['Symbol'] = symbol
            df['Returns'] = df['Close'].pct_change()
            df['Log_Returns'] = np.log(df['Close'] / df['Close'].shift(1))
            df.reset_index(inplace=True)
            all_data.append(df)
            
        except Exception as e:
            print(f"Error downloading {symbol}: {str(e)}")
    
    combined_df = pd.concat(all_data, ignore_index=True)
    return combined_df

# Exécution
symbols = ['AAPL', 'GOOGL', 'MSFT', 'NVDA', 'TSLA', 
           '^GSPC', '^VIX', 'BTC-USD']

data = download_financial_data(symbols)
data.to_csv('data/raw/financial_data_raw.csv', index=False)
data.to_pickle('data/raw/financial_data_raw.pkl')

print(data.info())
print(data.describe())